node/narwhal/src/helpers/codec.rs

```diff
@@ -84,3 +92,244 @@ impl<N: Network> Decoder for EventCodec<N> {
+
+    fn encode(&mut self, message_or_bytes: EventOrBytes<N>, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        let ciphertext = match self.noise_state {
+            NoiseState::Handshake(ref mut noise) => {
+                match message_or_bytes {
+                    // Don't allow message sending before the noise handshake has completed.
+                    EventOrBytes::Event(_) => unimplemented!(),
+                    EventOrBytes::Bytes(bytes) => {
+                        let mut buffer = [0u8; MAX_MESSAGE_LEN];
+                        let len = noise
+                            .write_message(&bytes, &mut buffer[..])
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        buffer[..len].into()
+                    }
+                }
+            }
+
+            NoiseState::PostHandshake(ref mut noise) => {
+                // Encode the message using the event codec.
+                let mut bytes = BytesMut::new();
+                match message_or_bytes {
+                    // Don't allow sending raw bytes after the noise handshake has completed.
+                    EventOrBytes::Bytes(_) => panic!("Unsupported post-handshake"),
+                    EventOrBytes::Event(event) => self.event_codec.encode(event, &mut bytes)?,
+                }
+
+                // Chunk the payload if necessary and encrypt with Noise.
+                //
+                // A Noise transport message is simply an AEAD ciphertext that is less than or
+                // equal to 65535 bytes in length, and that consists of an encrypted payload plus
+                // 16 bytes of authentication data.
+                //
+                // See: https://noiseprotocol.org/noise.html#the-handshakestate-object
+                let encrypted_chunks = bytes
+                    .par_chunks(MAX_MESSAGE_LEN - 16)
+                    .enumerate()
+                    .map(|(nonce_offset, plaintext_chunk)| {
+                        let mut buffer = vec![0u8; MAX_MESSAGE_LEN];
+                        let len = noise
+                            .state
+                            .write_message(noise.tx_nonce + nonce_offset as u64, plaintext_chunk, &mut buffer)
+                            .map_err(|e| Self::Error::new(io::ErrorKind::InvalidInput, e))?;
+
+                        buffer.truncate(len);
+
+                        Ok(buffer)
+                    })
+                    .collect::<io::Result<Vec<Vec<u8>>>>()?;
+
+                let mut buffer = BytesMut::with_capacity(encrypted_chunks.len());
+                for chunk in encrypted_chunks {
+                    buffer.extend_from_slice(&chunk);
+                    noise.tx_nonce += 1;
+                }
+
+                buffer
+            }
+        };
+
+        // Encode the resulting ciphertext using the length-delimited codec.
+        self.codec.encode(ciphertext.freeze(), dst)
```

Comment:
I missed this before, but I think we might instead want to have this call within the branches in order to write directly into the output buffer, without going through intermediate bytes

Reply:
As discussed, we shouldn't do this because of the lenght delimited codec, we'd be writing a `u32` per chunk to the stream.